text/0111-auto-resource-detection.md

```diff
@@ -0,0 +1,110 @@
+# Automatic Resource Detection
+
+Describes a mechanism to support auto-detection of resource information.
+
+## Motivation
+
+Resource information, i.e. attributes associated with the entity producing telemetry, can currently be supplied to tracer and meter providers or appended in custom exporters. In addition to this, it would be useful to have a mechanism to automatically detect resource information from the host (e.g. from an environment variable or from aws, gcp, etc metadata) and apply this to traces and metrics. This will in many cases prevent users from having to manually configure resource information.
+
+Note there are some existing implementations of this already in the SDKs (see [below](#prior-art-and-alternatives)), but nothing currently in the specification.
+
+## Explanation
+
+In order to apply auto-detected resource information to traces & metrics, a user will need to:
+
+- Configure which resource detector(s) they would like to run (e.g. AWS EC2 detector)
+- Provide the resource information returned by the configured detector(s) to the relevant tracer or meter provider(s)
+
+Note this means that if the user wants to make use of auto-detected resource information, they will be required to explicitly pass the resource to the trace and/or metric provider(s).
+
+If multiple detectors are configured, and more than one of these successfully detects a resource, the resources will be merged according to the Merge interface already defined in the specification, i.e. the earliest matched resource's attributes will take precedence. Each detector may be run in parallel, but to ensure deterministic results, the resources must be merged in the order the detectors were added.
+
+A default implementation of a detector that reads resource data from the `OTEL_RESOURCE` environment variable will be included in the SDK. The environment variable will contain of a list of key value pairs, and these are expected to be represented in a format similar to the [W3C Correlation-Context](https://github.com/w3c/correlation-context/blob/master/correlation_context/HTTP_HEADER_FORMAT.md#header-value), i.e.: `key1=value1,key2=value2`. This detector must always be configured as the first detector by default.
+
+Custom resource detectors related to specific environments (e.g. specific cloud vendors) must be implemented outside of the SDK, and users will need to import these separately.
```

Comment:
Resource detection is fairly trivial for all the major cloud platforms. You make an HTTP call to some well-known endpoint and get JSON or some plain text data. I don't think we necessarily need to use third party libraries which may bring in more dependencies than we care for. This would also simplify the number of dependencies for users as it'd all live in the SDK.

Reply:
Adding generic and configurable detectors as a package would be useful. Providing not only the detector interface, but also including configurable Environment Variable and HTTP detectors could be broadly useful.

For example,

```
type Resourcer interface {
	Resource(io.Reader) (*Resource, error)
}

type HTTPDetector struct {
	Client *http.Client
	Request http.Request
	Resourcer  Resourcer
}

func (d HTTPDetector) Detect(ctx context.Context) (*Resource, error) {
	resp, err := d.Client.Do(d.Request)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()
	return d.Resourcer.Resource(resp.Body)
}
```

Reply:
If we have default detectors, they should just be part of the SDK imo (as per the env variable detector mentioned in this OTEP). A generic HTTP detector could be useful as an implementation detail, but probably doesn't need to be mentioned in the spec.